src/riak_kv_w1c_worker.erl

```diff
@@ -222,34 +231,54 @@ async_put_replies(ReqIdTuples, Options) ->
+        warn_object_size = app_helper:get_env(riak_kv, warn_object_size)}}.
 
 handle_call(_Request, _From, State) ->
     {reply, undefined, State}.
 
-handle_cast({put, Bucket, BucketProps, PartitionIdx, Key, EncodedVal, ReqId, Preflist, #rec{from=From}=Rec}, #state{proxies=Proxies}=State) ->
-    NewState = case store_request_record(ReqId, Rec, State) of
-                   {undefined, S} ->
-                       UpdProxies = send_vnodes(Preflist, Proxies, Bucket, Key, EncodedVal, ReqId),
-                       postcommit(data_type_by_key(Key), Bucket,
-                                  maybe_ts_markup(PartitionIdx, Bucket, Key, EncodedVal),
-                                  BucketProps),
-                       S#state{proxies=UpdProxies};
-                   {_, S} ->
-                       reply(From, ReqId, {error, request_id_already_defined}),
-                       S
-    end,
+handle_cast({put, Bucket, BucketProps, PartitionIdx, Key, EncodedVal, ReqId, Preflist, Rec}, State) ->
+    #rec{from=From, obj_sizes=BinSizes} = Rec,
+    #state{proxies=Proxies, max_object_size=MaxSize, warn_object_size=WarnSize} = State,
+    %% Report or fail on large objects
+    NewState = case are_object_sizes_valid(Bucket, Key, BinSizes, MaxSize, WarnSize) of
+        true ->
+            case store_request_record(ReqId, Rec, State) of
+                {undefined, S} ->
+                    UpdProxies = send_vnodes(Preflist, Proxies, Bucket, Key, EncodedVal, ReqId),
+                    postcommit(data_type_by_key(Key), Bucket,
+                               maybe_ts_markup(PartitionIdx, Bucket, Key, EncodedVal),
+                               BucketProps),
+                    S#state{proxies=UpdProxies};
+                {_, S} ->
+                    reply(From, ReqId, {error, request_id_already_defined}),
+                    S
+            end;
+        false ->
+            reply(From, ReqId, {error, too_large}),
+            State
+     end,
     {noreply, NewState};
-handle_cast({batch_put, Bucket, BucketProps, EncodedVals, ReqId, PK, Preflist, #rec{from=From}=Rec}, #state{proxies=Proxies}=State) ->
-    NewState = case store_request_record(ReqId, Rec, State) of
-                   {undefined, S} ->
-                       UpdProxies = batch_send_vnodes(Preflist, Proxies, EncodedVals, ReqId),
-                       postcommit(ts, Bucket, {PK, EncodedVals}, BucketProps),
-                       S#state{proxies=UpdProxies};
-                   {_, S} ->
-                       reply(From, ReqId, {error, request_id_already_defined}),
-                       S
-    end,
+handle_cast({batch_put, Bucket, BucketProps, EncodedVals, ReqId, PK, Preflist, Rec}, State) ->
+    #rec{from=From, obj_sizes=BinSizes} = Rec,
+    #state{proxies=Proxies, max_object_size=MaxSize, warn_object_size=WarnSize} = State,
+    %% Report or fail on large objects
+    NewState = case are_object_sizes_valid(Bucket, PK, BinSizes, MaxSize, WarnSize) of
```

Comment:
This nested `case` statement falls afoul of our preferred Erlang style. I don't think it warrants a rewrite, but this function is getting a bit clunky.